system/ymodem/rb_main.c

```diff
@@ -0,0 +1,195 @@
+            {
+              sprintf(pathname, "%s%s", ym_fd->pathname, ctx->file_name);
+            }
+          else
+            {
+              sprintf(pathname, "%s/%s", ym_fd->pathname, ctx->file_name);
+            }
+
+          filename = pathname;
+        }
+      else
+        {
+          filename = ctx->file_name;
+        }
+
+      ym_fd->file_fd = open(filename, O_CREAT | O_RDWR);
+      if (ym_fd->file_fd < 0)
+        {
+          return -1;
+        }
+
+      ym_fd->file_saved_size = 0;
+    }
+  else if (ctx->packet_type == YMODEM_RECV_DATA_PKT)
+    {
+      if (ym_fd->file_saved_size + ctx->packet_size > ctx->file_length)
+        {
+          size = ctx->file_length - ym_fd->file_saved_size;
+        }
+      else
+        {
+          size = ctx->packet_size;
+        }
+
+      ret = write(ym_fd->file_fd, ctx->data, size);
+      if (ret != size)
+        {
+          return -1;
+        }
+
+      ym_fd->file_saved_size += ret;
+    }
+
+  return 0;
+}
+
+static void show_usage(FAR const char *progname, int errcode)
+{
+  fprintf(stderr, "USAGE: %s [OPTIONS]\n", progname);
+  fprintf(stderr, "\nWhere:\n");
+  fprintf(stderr, "\nand OPTIONS include the following:\n");
+  fprintf(stderr,
+          "\t-d <device>: Communication device to use.  Default: %s\n",
+          CONFIG_SYSTEM_YMODEM_DEVNAME);
+  fprintf(stderr,
+          "\t-p <path>: Save remote file path. Default: Current path\n");
+  exit(errcode);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+int main(int argc, FAR char *argv[])
+{
+  FAR const char *devname = CONFIG_SYSTEM_YMODEM_DEVNAME;
+  struct termios saveterm;
+  struct ymodem_ctx ctx;
+  struct termios term;
+  struct ym_fd ym_fd;
+  int option;
+
+  memset(&ym_fd, 0, sizeof(struct ym_fd));
+  while ((option = getopt(argc, argv, "d:p:h")) != ERROR)
+    {
+      switch (option)
+        {
+          case 'd':
+            devname = optarg;
+            break;
+          case 'h':
+            show_usage(argv[0], EXIT_FAILURE);
+          case 'p':
+            strncpy(ym_fd.pathname, optarg, PATH_MAX);
+            break;
+          default:
+          case '?':
+            fprintf(stderr, "ERROR: Unrecognized option\n");
+            show_usage(argv[0], EXIT_FAILURE);
+            break;
+        }
+    }
+
+  memset(&ctx, 0, sizeof(struct ymodem_ctx));
+  ctx.send = ymodem_uart_send;
+  ctx.receive = ymodem_uart_receive;
+  ctx.packet_handler = handler;
+  ctx.timeout = 200;
+  ctx.priv = &ym_fd;
+
+  ym_fd.file_fd = 0;
+  ym_fd.uart_fd = open(devname, O_RDWR | O_NONBLOCK);
+  if (ym_fd.uart_fd < 0)
+    {
+      printf("open\n");
+      return 0;
+    }
+
+  tcgetattr(ym_fd.uart_fd, &saveterm);
+  tcgetattr(ym_fd.uart_fd, &term);
+  cfmakeraw(&term);
+  tcsetattr(ym_fd.uart_fd, TCSANOW, &term);
+
+  sched_lock();
```

Comment:
why need sched_lock()?

Reply:
Make sure that other applications will not use/dev/console

Reply:
It's better not to use the kernel API, the application is better to keep POSIX compatible